lsh-2.1/doc/lsh-decode-key.1
.\" COPYRIGHT AND PERMISSION NOTICE
.\"
.\" Copyright (C) 1999 J.H.M. Dassen (Ray) <jdassen@wi.LeidenUniv.nl>
.\"
.\" Permission is granted to make and distribute verbatim copies of this
.\" manual provided the copyright notice and this permission notice are 
.\" preserved on all copies.
.\"
.\" Permission is granted to copy and distribute modified versions of this
.\" manual under the conditions for verbatim copying, provided that the
.\" entire resulting derived work is distributed under the terms of a 
.\" permission notice identical to this one.
.\"
.\" Permission is granted to copy and distribute translations of this manual
.\" into another language, under the above conditions for modified versions,
.\" except that this permission notice may be stated in a translation approved
.\" by the Free Software Foundation, Inc. <URL:http://www.fsf.org>
.\"
.\" END COPYRIGHT AND PERMISSION NOTICE
.\"
.\" If you make modified versions of this manual, please notify the current 
.\" maintainers of the package you received this manual from and make your
.\" modified versions available to them.
.\"
.TH LSH-DECODE-KEY 1 "NOVEMBER 2004" LSH-DECODE-KEY "Lsh Manuals"
.SH NAME
lsh-decode-key \- Convert a raw OpenSSH/SSH2 (IETF style) public key to sexp format.
.SH SYNOPSIS
.B lsh-decode-key
[\fIOPTION\fR...]

.SH DESCRIPTION
.B CAUTION! The information in this manpage may be invalid or outdated. For authorative
.B information on lsh, please see it's Texinfo manual (see the
.I SEE\ ALSO
.B section).

The public key (in OpenSSH/SSH2 format) is read from stdin,
converted to sexp format and written to stdout or another file specified.

This program is usually invoked by the ssh-conv script.

.SH OPTIONS

.TP
\fB\-b\fR, \fB\-\-base64\fR
Input is base64 encoded
.HP
\fB\-o\fR, \fB\-\-output\-file\fR=\fIFilename\fR Default is stdout
.TP
\fB\-\-debug\fR
Print huge amounts of debug information
.TP
\fB\-\-log\-file\fR=\fIFile\fR name
Append messages to this file.
.TP
\fB\-q\fR, \fB\-\-quiet\fR
Suppress all warnings and diagnostic messages
.TP
\fB\-\-trace\fR
Detailed trace
.TP
\fB\-v\fR, \fB\-\-verbose\fR
Verbose diagnostic messages
.TP
-?, \fB\-\-help\fR
Give this help list
.TP
\fB\-\-usage\fR
Give a short usage message
.TP
\fB\-V\fR, \fB\-\-version\fR
Print program version
.PP
Mandatory or optional arguments to long options are also mandatory or optional
for any corresponding short options.
.PP


.SH DIAGNOSTICS
See the 
.B --verbose
,
.B --trace
and 
.B --debug
options.

.SH "REPORTING BUGS"
Report bugs to <bug-lsh@gnu.org>.





.SH COPYING
The lsh suite of programs is distributed under the GNU General Public
License; see the COPYING and AUTHORS files in the source distribution for
details.
.SH AUTHOR
The lsh program suite is written mainly by Niels M\[:o]ller <nisse@lysator.liu.se>.

This man-page was written for lsh 2.0 by Pontus Freyhult
<pont_lsh@soua.net>, it is based on a template written by
J.H.M. Dassen (Ray) <jdassen@wi.LeidenUniv.nl>.

.SH "SEE ALSO"
.BR DSA (5),
.BR lsh (1),
.BR lsh-authorize (1),
.BR lsh-export-key (1),
.BR lsh-keygen (1),
.BR lsh-make-seed (1),
.BR lsh-upgrade (1),
.BR lsh-upgrade-key (1),
.BR lsh-writekey (1),
.BR lshd (8),
.BR secsh (5),
.BR SHA (5),
.BR SPKI (5),
.BR ssh-conv (1)

The full documentation for
.B lsh
is maintained as a Texinfo manual.  If the
.B info
and
.B lsh
programs are properly installed at your site, the command
.IP
.B info lsh
.PP
should give you access to the complete manual.


lsh-2.1/src/lsh-decode-key.c
/* lsh-decode-key.c
 *
 * Decode ssh2 keys.
 *
 */

#if HAVE_CONFIG_H
#include "config.h"
#endif

#include <fcntl.h>
#include <unistd.h>

#include <nettle/sexp.h>

#include "format.h"
#include "io.h"
#include "lsh_argp.h"
#include "lsh_string.h"
#include "publickey_crypto.h"
#include "spki.h"
#include "version.h"
#include "werror.h"
#include "xalloc.h"

#include "lsh-decode-key.c.x"

/* Option parsing */

const char *argp_program_version
= "lsh-decode-key-" VERSION;

const char *argp_program_bug_address = BUG_ADDRESS;

/* GABA:
   (class
     (name lsh_decode_key_options)
     (vars
       ; Output filename
       (file string)

       ; Assume input is base64
       (base64 . int)))
*/

static struct lsh_decode_key_options *
make_lsh_decode_key_options(void)
{
  NEW(lsh_decode_key_options, self);
  self->file = NULL;
  self->base64 = 0;

  return self;
}

static const struct argp_option
main_options[] =
{
  /* Name, key, arg-name, flags, doc, group */
  { "output-file", 'o', "Filename", 0, "Default is stdout", 0 },
  { "base64", 'b', NULL, 0, "Input is base64 encoded", 0 },
  { NULL, 0, NULL, 0, NULL, 0 }
};

static const struct argp_child
main_argp_children[] =
{
  { &werror_argp, 0, "", 0 },
  { NULL, 0, NULL, 0}
};

static error_t
main_argp_parser(int key, char *arg, struct argp_state *state)
{
  CAST(lsh_decode_key_options, self, state->input);

  switch(key)
    {
    default:
      return ARGP_ERR_UNKNOWN;

    case ARGP_KEY_END:
      break;
      
    case 'b':
      self->base64 = 1;
      break;
      
    case 'o':
      self->file = ssh_format("%lz", arg);
      break;
    }
  return 0;
}

static const struct argp
main_argp =
{ main_options, main_argp_parser, 
  NULL,
  ( "Converts a raw OpenSSH/ssh2 public key to sexp-format.\v"
    "Usually invoked by the ssh-conv script."),
  main_argp_children,
  NULL, NULL
};


static struct lsh_string *
lsh_decode_key(struct lsh_string *contents)
{
  struct simple_buffer buffer;
  int type;

  simple_buffer_init(&buffer, STRING_LD(contents));

  if (!parse_atom(&buffer, &type))
    {
      werror("Invalid (binary) input data.\n");
      return NULL;
    }

  switch (type)
    {
    case ATOM_SSH_DSS:
      {
        struct verifier *v;
        
        werror("Reading key of type ssh-dss...\n");

        v = parse_ssh_dss_public(&buffer);
        
        if (!v)
          {
            werror("Invalid dsa key.\n");
            return NULL;
          }
        else
          return PUBLIC_SPKI_KEY(v, 1);
      }
      
    case ATOM_SSH_RSA:
      {
          struct verifier *v;
          
          werror("Reading key of type ssh-rsa...\n");

          v = parse_ssh_rsa_public(&buffer);

          if (!v)
            {
              werror("Invalid rsa key.\n");
              return NULL;
            }
          else
            return PUBLIC_SPKI_KEY(v, 1);
      }      
    default:
      werror("Unknown key type.");
      return NULL;
    }
}


int main(int argc, char **argv)
{
  struct lsh_decode_key_options *options = make_lsh_decode_key_options();
  const struct exception *e;
  struct lsh_string *input;
  struct lsh_string *output;
  
  int out = STDOUT_FILENO;
  
  argp_parse(&main_argp, argc, argv, 0, NULL, options);

  if (options->file)
    {
      out = open(lsh_get_cstring(options->file),
                 O_WRONLY | O_CREAT, 0666);
      if (out < 0)
        {
          werror("Failed to open file `%S' for writing %e\n",
                 options->file, errno);
          return EXIT_FAILURE;
        }
    }

  input = io_read_file_raw(STDIN_FILENO, 3000);
  if (!input)
    {
      werror("Failed to read stdin %e\n", errno);
      return EXIT_FAILURE;
    }

  if (options->base64)
    {
      if (!lsh_string_base64_decode(input))
	{
          werror("Invalid base64 encoding.\n");

	  lsh_string_free(input);

          return EXIT_FAILURE;
        }
    }
  
  output = lsh_decode_key(input);
  lsh_string_free(input);

  if (!output)
    {
      werror("Invalid ssh2 key.\n");
      return EXIT_FAILURE;
    }
    
  e = write_raw(out, STRING_LD(output));
  lsh_string_free(output);
  
  if (e)
    {
      werror("Write failed: %z\n",
             e->msg);
      return EXIT_FAILURE;
    }
  
  gc_final();
  
  return EXIT_SUCCESS;
}
